src/solver/tron.jl

```diff
@@ -0,0 +1,414 @@
+              η₀ :: Real=1e-4,
+              η₁ :: Real=0.25,
+              η₂ :: Real=0.75,
+              Δmax :: Real=1e6
+             )
+  ℓ = nlp.meta.lvar
+  u = nlp.meta.uvar
+  f(x) = obj(nlp, x)
+  g(x) = grad(nlp, x)
+  n = nlp.meta.nvar
+
+  iter = 0
+  start_time = time()
+  el_time = 0.0
+
+  # Preallocation
+  temp = zeros(n)
+  gpx = zeros(n)
+  xc = zeros(n)
+  gc = zeros(n)
+  Hs = zeros(n)
+
+  x = max.(ℓ, min.(nlp.meta.x0, u))
+  fx = f(x)
+  gx = g(x)
+  num_success_iters = 0
+
+  # Optimality measure
+  project_step!(gpx, x, -gx, ℓ, u)
+  πx = norm(gpx)
+  ϵ = atol + rtol * πx
+  optimal = πx <= ϵ
+  tired = iter >= itmax || el_time > timemax
+  unbounded = fx < fmin
+  stalled = false
+
+  αC = 1.0
+  Δ = min(max(1.0, 0.1*norm(πx)), 100)
+  if verbose
+    @printf("%4s  %9s  %7s  %7s  %7s  %s\n", "Iter", "f", "π", "Radius", "Ratio", "CG-status")
+    @printf("%4d  %9.2e  %7.1e  %7.1e\n", iter, fx, πx, Δ)
+  end
+  while !(optimal || tired || stalled)
+    # Current iteration
+    xc .= x
+    fc = fx
+    gc .= gx
+
+    # Model
+    H = hess_op!(nlp, x, temp)
+
+    # Cauchy starts here
+    αC, s = cauchy(x, H, gx, Δ, αC, ℓ, u, μ₀=μ₀, μ₁=μ₁, σ=σ)
+
+    s, cgits, cginfo = projected_newton!(x, H, gx, Δ, cgtol, s, ℓ, u, max_cgiter=max_cgiter)
+    slope, qs = compute_Hs_slope_qs!(Hs, H, s, gx)
+    Pred = -qs
+
+    fx = f(x)
+    Ared = fc - fx
+    if num_success_iters == 0
+      Δ = min(Δ, norm(s))
+    end
+
+    # Update the trust region
+    γ = fx - fc - slope
+    if γ <= 0.0
+      α = σ₃
+    else
+      α = max(σ₁, -0.5*slope/γ)
+    end
+
+    success = Ared >= η₀ * Pred
+
+    if Ared < η₀ * Pred
+      Δ = min(max(α, σ₁) * norm(s), σ₂ * Δ)
+    elseif Ared < η₁ * Pred
+      Δ = max(σ₁ * Δ, min(α * norm(s), σ₂ * Δ))
+    elseif Ared < η₂ * Pred
+      Δ = max(σ₁ * Δ, min(α * norm(s), σ₃ * Δ))
+    else
+      Δ = max(Δ, min(α * norm(s), σ₃ * Δ))
+    end
+
+    if success
+      num_success_iters += 1
+      gx = g(x)
+      project_step!(gpx, x, -gx, ℓ, u)
+      πx = norm(gpx)
+    end
+
+    # No post-iteration
+
+    if !success
+      fx = fc
+      x .= xc
+    end
+
+    iter += 1
+    el_time = time() - start_time
+    tired = iter >= itmax ||  el_time > timemax
+    optimal = πx <= ϵ || fx == -Inf
+
+    verbose && @printf("%4d  %9.2e  %7.1e  %7.1e  %7.1e  %s\n", iter, fx, πx, Δ, Ared/Pred, cginfo)
+  end
+
+  status = if tired
+    iter >= itmax ? "maximum number of iterations" : "maximum elapsed time"
+  elseif πx <= ϵ
+    "first-order stationary"
+  elseif fx == -Inf
+    "objective function is unbounded from below"
+  elseif stalled
+    "stalled"
+  else
+    "unknown"
+  end
+
+  return x, fx, πx, iter, optimal, tired, status, el_time
+end
+
+"""
+    nbrk, brkmin, brkmax = breakpoints(x, d, ℓ, u)
+
+Find the smallest and largest values of `α` such that `x + αd` lies on the
+boundary. `x` is assumed to be feasible. `nbrk` is the number of violated
+bounds in the given direction.
```

Comment:
"the number of breakpoints from `x` in the direction `d`" would be more accurate I believe.